p4-16/spec/P4-16-spec.adoc

```diff
@@ -3581,6 +3581,12 @@ Bit-strings also support the following operations:
   The effect of this statement is to set bits `H` through `L` (inclusive of
   both) of `e` to the bit-pattern represented by `x`, and leaves all other bits
   of `e` unchanged.  A slice of an unsigned integer is an unsigned integer.
+  A slice may also be specified as `[L+:W]` where `L` is the lowest bit of the slice
+  and `W` is the width of the slice.  In this case, only `W` must be a non-negative
+  local compile-time known numeric value.  `L` must also be a numeric value and
+  some architectures may require it to be compile-time known.  It must also be
+  non-negative and in-range for the type.  Some architectures may allow variable indexing,
+  which might prevent diagnosing an out of range value.
```

Comment:
We should define semantics of out-of-bounds slices. Similar to how we have it defined for header stack indexing, we should not leave it to the compiler.

- we might consider the "out-of-range" bits to be undefined
- or we might want then to be 0s
- we probably want write to out-of-bounds bits to have no effect.

I would personally prefer the read values to be zeroed, e.g.

```p4
bit<42> foo = ~0;
bit<4> bar = foo[a+:4];
```

`bar` would be

- `0b111` for `a = 0`, `a = 10`, etc.
- `0b0011` for `a = 40`
- `0b0000` for `a > 42`

I.e. this would be equivalent to `(bit<4>)(foo >> a)`.

Reply:
This should have the same semantics as shift.